File: twidere/src/main/java/org/mariotaku/twidere/activity/support/FileSelectorActivity.java
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.activity.support;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.DialogFragment;

import org.mariotaku.twidere.fragment.support.FileSelectorDialogFragment;
import org.mariotaku.twidere.util.PermissionUtils;
import org.mariotaku.twidere.util.ThemeUtils;

import java.io.File;

import static android.os.Environment.getExternalStorageDirectory;

public class FileSelectorActivity extends BaseSupportDialogActivity implements FileSelectorDialogFragment.Callback {


    private Runnable mResumeFragmentsRunnable;

    @Override
    public int getThemeResourceId() {
        return ThemeUtils.getNoDisplayThemeResource(this);
    }

    @Override
    public void onCancelled(final DialogFragment df) {
        if (!isFinishing()) {
            finish();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        setVisible(true);
    }

    @Override
    public void onDismissed(final DialogFragment df) {
        if (!isFinishing()) {
            finish();
        }
    }

    @Override
    public void onFilePicked(final File file) {
        final Intent intent = new Intent();
        intent.setData(Uri.fromFile(file));
        setResult(RESULT_OK, intent);
        finish();
    }

    @Override
    protected void onResumeFragments() {
        super.onResumeFragments();
        if (mResumeFragmentsRunnable != null) {
            mResumeFragmentsRunnable.run();
        }
    }

    @Override
    public void onRequestPermissionsResult(final int requestCode, final @NonNull String[] permissions,
                                           final @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_REQUEST_PERMISSIONS) {
            mResumeFragmentsRunnable = new Runnable() {
                @Override
                public void run() {
                    if (PermissionUtils.getPermission(permissions, grantResults, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED &&
                            PermissionUtils.getPermission(permissions, grantResults, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                        showPickFileDialog();
                    } else {
                        finishWithDeniedMessage();
                    }
                }
            };
        }
    }

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final Intent intent = getIntent();

        final String action = intent.getAction();
        if (!INTENT_ACTION_PICK_FILE.equals(action) && !INTENT_ACTION_PICK_DIRECTORY.equals(action)) {
            finish();
            return;
        }
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
            showPickFileDialog();
        } else if (Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN) {
            final String[] permissions = {Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE};
            ActivityCompat.requestPermissions(this, permissions, REQUEST_REQUEST_PERMISSIONS);
        } else {
            finishWithDeniedMessage();
        }
    }

    private void finishWithDeniedMessage() {
        if (isFinishing()) return;
        finish();
    }

    private void showPickFileDialog() {
        final Intent intent = getIntent();
        final Uri data = intent.getData();
        final String action = intent.getAction();
        File initialDirectory = data != null ? new File(data.getPath()) : getExternalStorageDirectory();
        if (initialDirectory == null) {
            initialDirectory = new File("/");
        }
        final FileSelectorDialogFragment f = new FileSelectorDialogFragment();
        final Bundle args = new Bundle();
        args.putString(EXTRA_ACTION, action);
        args.putString(EXTRA_PATH, initialDirectory.getAbsolutePath());
        args.putStringArray(EXTRA_FILE_EXTENSIONS, intent.getStringArrayExtra(EXTRA_FILE_EXTENSIONS));
        f.setArguments(args);
        f.show(getSupportFragmentManager(), "select_file");
    }

}

File: twidere/src/main/kotlin/org/mariotaku/twidere/activity/FileSelectorActivity.kt
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.activity

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment.getExternalStorageDirectory
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.fragment.app.DialogFragment
import org.mariotaku.ktextension.Bundle
import org.mariotaku.ktextension.checkAllSelfPermissionsGranted
import org.mariotaku.ktextension.set
import org.mariotaku.twidere.R
import org.mariotaku.twidere.TwidereConstants.REQUEST_REQUEST_PERMISSIONS
import org.mariotaku.twidere.constant.IntentConstants.*
import org.mariotaku.twidere.fragment.FileSelectorDialogFragment
import java.io.File
import android.Manifest.permission as AndroidPermissions

class FileSelectorActivity : BaseActivity(), FileSelectorDialogFragment.Callback {

    private val PICKER_REQUEST_CODE: Int = 54837

    override fun onCancelled(df: DialogFragment) {
        if (!isFinishing) {
            finish()
        }
    }

    override fun onStart() {
        super.onStart()
        setVisible(true)
    }

    override fun onDismissed(df: DialogFragment) {
        if (!isFinishing) {
            finish()
        }
    }

    override fun onFilePicked(file: File) {
        val intent = Intent()
        intent.data = Uri.fromFile(file)
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_REQUEST_PERMISSIONS) {
            executeAfterFragmentResumed {
                if (checkAllSelfPermissionsGranted(AndroidPermissions.READ_EXTERNAL_STORAGE, AndroidPermissions.WRITE_EXTERNAL_STORAGE)) {
                    showPickFileDialog()
                } else {
                    finishWithDeniedMessage()
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val action = intent.action
        if (INTENT_ACTION_PICK_FILE != action && INTENT_ACTION_PICK_DIRECTORY != action) {
            finish()
            return
        }
        when {
            checkAllSelfPermissionsGranted(AndroidPermissions.READ_EXTERNAL_STORAGE, AndroidPermissions.WRITE_EXTERNAL_STORAGE) -> {
                showPickFileDialog()
            }
            Build.VERSION.SDK_INT > Build.VERSION_CODES.JELLY_BEAN -> {
                val permissions = arrayOf(AndroidPermissions.READ_EXTERNAL_STORAGE, AndroidPermissions.WRITE_EXTERNAL_STORAGE)
                ActivityCompat.requestPermissions(this, permissions, REQUEST_REQUEST_PERMISSIONS)
            }
            else -> {
                finishWithDeniedMessage()
            }
        }
    }

    private fun finishWithDeniedMessage() {
        if (isFinishing) return
        Toast.makeText(this, R.string.message_toast_select_file_no_storage_permission, Toast.LENGTH_SHORT).show()
        finish()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == PICKER_REQUEST_CODE && resultCode == Activity.RESULT_OK && data != null) {
            setResult(Activity.RESULT_OK, Intent().also { it.data = data.data })
            finish()
        } else {
            if (!isFinishing) {
                finish()
            }
        }
    }

    private fun showPickFileDialog() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            Intent().apply {
                if (intent.action == INTENT_ACTION_PICK_FILE) {
                    action = Intent.ACTION_GET_CONTENT
                    type = "*/*"
                } else if (intent.action == INTENT_ACTION_PICK_DIRECTORY) {
                    action = Intent.ACTION_OPEN_DOCUMENT_TREE
                }
            }.also {
                startActivityForResult(
                        Intent.createChooser(it, getString(R.string.pick_file)),
                        PICKER_REQUEST_CODE
                )
            }
        } else {
            val initialDirectory = intent?.data?.path?.let(::File) ?: getExternalStorageDirectory()
            ?: File("/")
            val f = FileSelectorDialogFragment()
            f.arguments = Bundle {
                this[EXTRA_ACTION] = intent.action
                this[EXTRA_PATH] = initialDirectory.absolutePath
                this[EXTRA_FILE_EXTENSIONS] = intent.getStringArrayExtra(EXTRA_FILE_EXTENSIONS)
            }
            f.show(supportFragmentManager, "select_file")
        }
    }

}
